contracts/src/v0.8/vrf/dev/VRFCoordinatorV2_5.sol

```diff
@@ -379,123 +390,140 @@ contract VRFCoordinatorV2_5 is VRF, SubscriptionAPI, IVRFCoordinatorV2Plus {
+    _chargePayment(payment, nativePayment, rc.subId);
 
-      return payment;
+    // Include payment in the event for tracking costs.
+    emit RandomWordsFulfilled(output.requestId, output.randomness, rc.subId, payment, success, onlyPremium);
+
+    return payment;
+  }
+
+  function _chargePayment(uint96 payment, bool nativePayment, uint256 subId) internal {
+    if (nativePayment) {
+      if (s_subscriptions[subId].nativeBalance < payment) {
+        revert InsufficientBalance();
+      }
+      s_subscriptions[subId].nativeBalance -= payment;
+      s_withdrawableNative += payment;
+    } else {
+      if (s_subscriptions[subId].balance < payment) {
+        revert InsufficientBalance();
+      }
+      s_subscriptions[subId].balance -= payment;
+      s_withdrawableTokens += payment;
     }
   }
 
   function _calculatePaymentAmount(
     uint256 startGas,
-    uint256 gasAfterPaymentCalculation,
     uint256 weiPerUnitGas,
-    bool nativePayment
-  ) internal view returns (uint96) {
+    bool nativePayment,
+    bool onlyPremium
+  ) internal returns (uint96) {
     if (nativePayment) {
-      return
-        _calculatePaymentAmountNative(
-          startGas,
-          gasAfterPaymentCalculation,
-          s_feeConfig.fulfillmentFlatFeeNativePPM,
-          weiPerUnitGas
-        );
+      return _calculatePaymentAmountNative(startGas, weiPerUnitGas, onlyPremium);
     }
-    return
-      _calculatePaymentAmountLink(
-        startGas,
-        gasAfterPaymentCalculation,
-        s_feeConfig.fulfillmentFlatFeeLinkPPM,
-        weiPerUnitGas
-      );
+    return _calculatePaymentAmountLink(startGas, weiPerUnitGas, onlyPremium);
   }
 
   function _calculatePaymentAmountNative(
     uint256 startGas,
-    uint256 gasAfterPaymentCalculation,
-    uint32 fulfillmentFlatFeePPM,
-    uint256 weiPerUnitGas
-  ) internal view returns (uint96) {
+    uint256 weiPerUnitGas,
+    bool onlyPremium
+  ) internal returns (uint96) {
     // Will return non-zero on chains that have this enabled
     uint256 l1CostWei = ChainSpecificUtil._getCurrentTxL1GasFees(msg.data);
     // calculate the payment without the premium
-    uint256 baseFeeWei = weiPerUnitGas * (gasAfterPaymentCalculation + startGas - gasleft());
-    // calculate the flat fee in wei
-    uint256 flatFeeWei = 1e12 * uint256(fulfillmentFlatFeePPM);
-    // return the final fee with the flat fee and l1 cost (if applicable) added
-    return uint96(baseFeeWei + flatFeeWei + l1CostWei);
+    uint256 baseFeeWei = weiPerUnitGas * (s_config.gasAfterPaymentCalculation + startGas - gasleft());
+    // calculate flat fee in native
+    uint256 flatFeeWei = 1e12 * uint256(s_config.fulfillmentFlatFeeNativePPM);
+    if (onlyPremium) {
+      return uint96((((l1CostWei + baseFeeWei) * (s_config.nativePremiumPercentage)) / 100) + flatFeeWei);
+    } else {
+      return uint96((((l1CostWei + baseFeeWei) * (100 + s_config.nativePremiumPercentage)) / 100) + flatFeeWei);
+    }
```

Comment:
Can save some bytecode

```
uint8 nativePercentage = s_config.nativePremiumPercentage;
if (!onlyPremium) {
  nativePercentage += 100;
}
return uint96((((l1CostWei + baseFeeWei) * nativePercentage) / 100) + flatFeeWei);
```

Reply:
will result in higher gas cost. leaving as is